mm2src/common/jsonrpc_client.rs

```diff
@@ -108,60 +238,145 @@ pub enum JsonRpcErrorType {
 pub trait JsonRpcMultiClient: JsonRpcClient {
-    fn transport_exact(&self, to_addr: String, request: JsonRpcRequest) -> JsonRpcResponseFut;
+    /// Sends the given `request` to the specified `to_addr` remote.
+    /// Returns either an address `JsonRpcRemoteAddr` of the responder and the `JsonRpcResponseEnum` response,
+    /// or a stringified error.
+    fn transport_exact(&self, to_addr: String, request: JsonRpcRequestEnum) -> JsonRpcResponseFut;
 
+    /// Sends the given single `request` to the specified `to_addr` remote and tries to decode the response into `T`.
     fn send_request_to<T: DeserializeOwned + Send + 'static>(
         &self,
         to_addr: &str,
         request: JsonRpcRequest,
     ) -> RpcRes<T> {
         let client_info = self.client_info();
         Box::new(
-            self.transport_exact(to_addr.to_owned(), request.clone())
-                .then(move |result| process_transport_result(result, client_info, request)),
+            self.transport_exact(to_addr.to_owned(), JsonRpcRequestEnum::Single(request.clone()))
+                .then(move |result| process_transport_single_result(result, client_info, request)),
         )
     }
 }
 
-fn process_transport_result<T: DeserializeOwned + Send + 'static>(
-    result: Result<(JsonRpcRemoteAddr, JsonRpcResponse), String>,
+/// Checks if the given `result` is success and contains `JsonRpcResponse`.
+/// Tries to decode the batch response into `T`.
+fn process_transport_single_result<T: DeserializeOwned + Send + 'static>(
+    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), String>,
     client_info: String,
     request: JsonRpcRequest,
 ) -> Result<T, JsonRpcError> {
-    let (remote_addr, response) = match result {
-        Ok(r) => r,
+    let request = JsonRpcRequestEnum::Single(request);
+
+    match result {
+        Ok((remote_addr, JsonRpcResponseEnum::Single(single))) => {
+            process_single_response(client_info, remote_addr, request, single)
+        },
+        Ok((remote_addr, JsonRpcResponseEnum::Batch(batch))) => {
+            let error = ERRL!("Expeced single response, found batch response: {:?}", batch);
+            Err(JsonRpcError {
+                client_info,
+                request,
+                error: JsonRpcErrorType::Parse(remote_addr, error),
+            })
+        },
+        Err(e) => Err(JsonRpcError {
+            client_info,
+            request,
+            error: JsonRpcErrorType::Transport(e),
+        }),
+    }
+}
+
+/// Checks if the given `result` is success and contains `JsonRpcBatchResponse`.
+/// Tries to decode the batch response into `Vec<T>` in the same order in which they were requested.
+fn process_transport_batch_result<T: DeserializeOwned + Send + 'static>(
+    result: Result<(JsonRpcRemoteAddr, JsonRpcResponseEnum), String>,
+    client_info: String,
+    request: JsonRpcBatchRequest,
+) -> Result<Vec<T>, JsonRpcError> {
+    let orig_ids: Vec<_> = request.orig_sequence_ids().collect();
```

Comment:
`request: JsonRpcBatchRequest` is moved when it's converted into `JsonRpcRequestEnum`:
https://github.com/KomodoPlatform/atomicDEX-API/blob/7224b85e39a2406b5b9e8cb9a8e69cd59fda514f/mm2src/common/jsonrpc_client.rs#L367-L368
I can't move these two lines below since `request: JsonRpcRequestEnum` is used to return an error